src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -86,5 +107,202 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+
+                    targetsList = lockFile?.Targets;
+                }
+
+                _lastTimeAssetsModified = fileInfo.LastWriteTimeUtc;
+                _lastTargets = targetsList;
+                _lastPackgeSpec = new WeakReference<PackageSpec>(currentPackageSpec);
+            }
+            else if (cacheHitTargets && cacheHitPackageSpec && lastPackageSpec != null && _lastTargets != null)
+            {
+                targetsList = _lastTargets;
+                isCacheHit = true;
+            }
+
+            return new RestoreGraphRead(currentPackageSpec, targetsList?.ToArray(), isCacheHit);
+        }
+
+        internal async ValueTask<TransitiveEntry> GetTransitivePackageOriginAsync(PackageIdentity transitivePackage, CancellationToken ct)
+        {
+            /** Pseudocode
+            1. Get project restore graph
+
+            2. If it is cached
+               2.1 Look for a transive cached entry
+               2.2 If found, return that entry
+
+            Otherwise:
+
+            3. For each target framework graph:
+              3.1 Foreach direct dependency d:
+                  3.1.1 do DFS to look for transitive dependency
+
+                  3.1.2 if found:
+                    Add to list, indexed by framework
+
+            4. Cache the result list and return it
+            */
+
+            ct.ThrowIfCancellationRequested();
+
+            RestoreGraphRead reading = await GetFullRestoreGraphAsync(ct);
+            if (reading.IsCacheHit)
+            {
+                // Assets file has not changed, look at transtive origin cache
+                var cacheEntry = GetCachedTransitiveOrigin(transitivePackage);
+                if (cacheEntry != null)
+                {
+                    return cacheEntry;
+                }
+            }
+            else
+            {
+                // Assets file changed, recompute transitive origins
+                CleanCache();
+            }
+
+            // Otherwise, find Transitive origin and update cache
+            var packageOrigins = new Dictionary<Tuple<NuGetFramework, string>, IReadOnlyList<PackageReference>>();
+
+            if (reading.TargetsList != null)
+            {
+                var pkgs = await GetInstalledAndTransitivePackagesAsync(ct);
+
+                var memory = new Dictionary<PackageIdentity, bool?>();
+
+                foreach (var targetFxGraph in reading.TargetsList)
+                {
+                    var key = Tuple.Create(targetFxGraph.TargetFramework, targetFxGraph.RuntimeIdentifier);
+                    var list = new List<PackageReference>();
+                    memory.Clear();
+
+                    foreach (var directPkg in pkgs.InstalledPackages) // InstalledPackages are direct dependencies
+                    {
+                        var found = FindTransitive(directPkg.PackageIdentity, transitivePackage, targetFxGraph, memory);
```

Comment:
consider pre-calculating *all* of the direct/transtive dependencies mapping. 

The graph walk to get one seems to be fairly equivalent to getting them all. 
It should be something that you can measure fairly easily. 

If you do that, you won't need to keep the `LockFileTarget` in memory, which is pretty large.

Reply:
How do I avoid blocking the caller when all direct/transitives mapping are calculated? Ideally, you just want one and others should be calculated in the background.

Reply:
You use TaskCompletionSource for something like that, but I'm wondering if it's an overkill.

Isn't calculating the transitive origins for a package *always* an exhaustive graph walk?
It's memory footprint that slightly changes depending whether you're trying to keep track of one or all packages.

Reply:
The API is intended to be called once for each transitive dependency.

There will be a tooltip binding on each transitive dependency, displayed in Package Search Results in PM UI.

It's possible to get all transitive dependencies if user see all tooltips.

Reply:
> Isn't calculating the transitive origins for a package always an exhaustive graph walk?

Yes, you need to walk the entire graph looking for a transitive origin. That's why I implemented some prunning on the DFS.

Reply:
> It's memory footprint that slightly changes depending whether you're trying to keep track of one or all packages.

Yeap, memory usage is roughly the same if we assume no changes in project.

Reply:
I believe what @Nikolche means is we do DFS on whole graph starting from direct dependencies only once to create all direct -> transitive map. Then we can reverse that relationship to get who is direct package brought in each transitive dependency.

Reply:
Here is mock code idea I wrote on notepad. Please note it's not complete, just to give idea.
Once BuildMapping creates dictionary then getting value from it would be O(1)

```

private Dictionary<> BuildMapping()
{
    RestoreGraphRead reading = await GetFullRestoreGraphAsync(ct);

    // Otherwise, find Transitive origin and update cache
    var packageOrigin = new Dictionary<Tuple<NuGetFramework, string>, List<PackageReference>>();
    var packageDependency = new Dictionary<PackageDependency, List<PackageReference>>();
    
    foreach (var targetFxGraph in reading.TargetsList)
    {
        var key = Tuple.Create(targetFxGraph.TargetFramework, targetFxGraph.RuntimeIdentifier);
        var list = new List<PackageReference>(); // Could be HashSet

        foreach (var directPkg in pkgs.InstalledPackages) // InstalledPackages are direct dependencies
        {
            GetTransitive(directPkg, directPkg, targetFxGraph, packageDependency);
        }
    }

    Now revert the packageDependency key-value relationship in dictionary to get packageOrigin.
           return packageOrigin;
}

private void GetTransitive(PackageIdentity top,PackageIdentity dependentPkg, LockFileTarget graph, Dictionary packageDependency)
{
    LockFileTargetLibrary node = graph
        .Libraries
        .Where(x => x.Name == current.Id && x.Version.Equals(current.Version) && x.Type == "package")
        .FirstOrDefault();

    if (node != default)
    {
        if(!packageDependency.ContainsKey(top))
        {
            packageDependency[top] = new List<>();
        }

        foreach (PackageDependency dep in node.Dependencies)
        {
            var pkgChild = new PackageIdentity(dep.Id, dep.VersionRange.MinVersion);
            packageDependency[top].Add(pkgChild);
            GetTransitive(top, pkgChild , graph, packageDependency);
        }
    }
}

```

Reply:
Please let me know if you have any question, I would happy to answer it to complete the idea.